File: ts/2025/day11.ts
let p1 = 0, p2 = 0
const G = new Map<string, string[]>()
const lines: string[] = (await Bun.stdin.text()).trim().split("\n")
for (const line of lines) {
	const [src, rest] = line.split(":")
	if (src !== undefined && rest !== undefined) {
		G.set(src, rest.split(" ").filter(Boolean))
	}
}

const f = (x: string): number => x === "out" ? 1 : (G.get(x) ?? []).reduce((m, n) => m + f(n), 0)
p1 = f("you")

const cache = new Map<string, number>()
const g = (x: string, fft: boolean, dac: boolean): number => {
	let res = cache.get(JSON.stringify([x, fft, dac]))
	if (res !== undefined) {
		return res
	}
	res = ((x: string, fft: boolean, dac: boolean): number => {
		if (x === "out") {
			return (fft && dac) ? 1 : 0
		}
		return (G.get(x) ?? []).reduce((m, n) => m + g(n, fft || n === "fft", dac || n === "dac"), 0)
	})(x, fft, dac)
	cache.set(JSON.stringify([x, fft, dac]), res)
	return res
}
p2 = g("svr", false, false)

console.log("Part 1: %d", p1)
console.log("Part 2: %d", p2)

File: py/2025/day11.py
import sys
from functools import cache

p1, p2 = 0, 0
D = sys.stdin.read()
G = {}
lines = [line for line in D.split("\n") if line]
for line in lines:
    src, rest = line.split(":")
    G[src] = rest.split()


def f(x: str) -> int:
    if x == "out":
        return 1
    else:
        return sum(f(node) for node in G.get(x, []))


p1 = f("you")


@cache
def g(x: str, fft: bool, dac: bool) -> int:
    if x == "out":
        return 1 if fft and dac else 0
    else:
        return sum(
            g(node, node == "fft" or fft, node == "dac" or dac) for node in G.get(x, [])
        )


p2 = g("svr", False, False)

print(f"The result for part 1: {p1}")
print(f"The result for part 2: {p2}")
